### PR TITLE
Avoid strcpy of Field Names in C-Ext Read

### PR DESCRIPTION
**Pull request by rmarrowstone**
This change removes a strcpy to a temporary stack allocated array for
the field name of a Struct field. Instead it just allocates the
PyUnicode object for the field name, which does a copy anyway. The
copy is important because the ION_STRING is just a pointer to the
buffer which will change as we descend into parsing the Struct.

The performance impact of this change is measurable for some datasets
but not terribly significant. It is nonetheless an apparent simplifier
and therefore worth doing regardless.

** Please double check my DECREF-ing.**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: amazon/ion/ioncmodule.c [hunk 1]
```diff
@@ -1057,22 +1055,11 @@ iERR ionc_read_value(hREADER hreader, ION_TYPE t, PyObject* container, BOOL in_s
...
-
-    char    field_name_value[FIELD_NAME_MAX_LEN];
-    int     field_name_len = 0;
-    BOOL    None_field_name = TRUE;
+    PyObject*   py_field_name = NULL;
 
     if (in_struct) {
         IONCHECK(ion_reader_get_field_name(hreader, &field_name));
-        field_name_len = field_name.length;
-        if (field_name_len > FIELD_NAME_MAX_LEN) {
```
Comment (nirosys) [inline amazon/ion/ioncmodule.c:1068]: oh yea, nice to see something like this go away.

File: amazon/ion/ioncmodule.c [hunk 2]
```diff
@@ -1057,22 +1055,11 @@ iERR ionc_read_value(hREADER hreader, ION_TYPE t, PyObject* container, BOOL in_s
...
-            _FAILWITHMSG(IERR_INVALID_ARG,
-                "Filed name overflow, please try again with pure python.");
-        }
-        if (field_name.value != NULL) {
-            None_field_name = FALSE;
-            strcpy(field_name_value, field_name.value);
```
Comment (nirosys) [inline amazon/ion/ioncmodule.c:1074]: Surprised this wasn't an issue previously, ion-c documents ION_STRINGs as not being null-terminated.

Comment (rmarrowstone) [reply amazon/ion/ioncmodule.c:1074]: Hah. It apparently was!
```
>>> from amazon.ion import simpleion
>>> simpleion.loads("{'f\0o': bar}")
[1]    18140 segmentation fault  python3
```

Works Now:
```
>>> from amazon.ion import simpleion
>>> simpleion.loads("{'f\0o': bar}")
{'f\x00o': IonPySymbol(text='bar', sid=None, location=None)}
```

Comment (rmarrowstone) [reply amazon/ion/ioncmodule.c:1074]: Created https://github.com/amazon-ion/ion-tests/pull/84 to add a test for this.

**Review by tgregg — APPROVED** [review]
The DECREF usage is correct as far as I can tell.

**Pull request merged**